---
title: "200204_HiSeq Data Analysis"
author: "Carmen Emborski"
date: "2/04/2020"
output: html_document
---

```{r setup, include=FALSE, cache = F}
knitr::opts_chunk$set(echo = TRUE, error = TRUE)
library(knitr)
library(kableExtra)
library(gridExtra)
library(scales)
library(RColorBrewer)
library(tidyverse) #dplyr package included
library(stringr)
library(lme4)
library(nlme)
library(effects)
library(DESeq2)
library(qiime2R)
library(phyloseq)
library(DivNet)
library(vegan)
library(sjPlot) #  mixed effects table summary
library(ggplot2)
library(devtools)
theme_set(theme_minimal() + theme(legend.position = "bottom", axis.title =  element_text(size = 20), axis.text =  element_text(size = 10), strip.text = element_text(size = 20)) + theme_bw())
```

##Import Files  
1) metadata file (provides 16s sample details (i.e. treatment, generations))  
2) feature table (this is the sequencing output file after qiime processing)  
3) taxonomy file (gives taxanomic names to unique amplicon sequence variants)  
4) tree (helps construct the phylogenetic tree)
```{r}
taxonomy<- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16s%20rRNA%20HiSeq%20Analysis/Taxonomy.tsv") 
names(taxonomy) <- c("row", "tax", "Confidence")
row.names(taxonomy) <-taxonomy[[1]]
taxonomy <- taxonomy[,(-1)]
taxonomy <-  separate(taxonomy, tax, c("D0","D1", "D2", "D3", "D4", "D5", "D6", "D7", "D8", "D9", "D10", "D11", "D12", "D13", "D14"), sep = ";", fill = "right")
taxonomy <- taxonomy[,c(1:6)]
taxmat <- as.matrix(taxonomy)
TAX = tax_table(taxmat) 

#import metadata
metatable <- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16s%20rRNA%20HiSeq%20Analysis/Metatable.txt")
row.names(metatable) <- metatable[[1]]
metatable<- metatable[,(-1)]
META <- sample_data(metatable)

#Import Phylogenetic tree (rooted, exported from qiime2)
TREE <- read_tree("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16s%20rRNA%20HiSeq%20Analysis/tree.nwk")

#Import feature (sequence variant) table
Svtab <- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16s%20rRNA%20HiSeq%20Analysis/FeatureTable.txt")
row.names(Svtab)<-Svtab[[1]] #make OTU ID the row names
Svtab<-Svtab[,-(1)] # remove OTU ID column
fmat <- as.matrix(Svtab)
OTU = otu_table(fmat, taxa_are_rows = TRUE)

#create phyloseq object
ps <- phyloseq(OTU, TAX, META, TREE) 
```
  
  
##Visualize Raw Data:
Generate a prevelance table (number of samples each taxa occurs in) for each taxa.
```{r Prevalence Plot Unfiltered}
ps <- subset_samples(ps, Treatment != "HSD")
ps

prevelancedf = apply(X = otu_table(ps),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
#Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(ps),
                      tax_table(ps))

#Phylum Level Prevalence Table
plyr::ddply(prevelancedf, "D1", function(df1){
  data.frame(mean_prevalence=mean(df1$Prevalence),total_abundance=sum(df1$TotalAbundance,na.rm = T),stringsAsFactors = F)
  })
```
  
  
##Preprocessing and Filtering:
###Prefilter 
Remove ambiguous phyla (NA), keep only bacterial OTUs, and remove taxa with less than 1 read (singletons) and those that cannot be part of the microbiome
```{r Prefilter}
ps0 = subset_taxa(ps, D0=="D_0__Bacteria")
ps0 <- subset_taxa(ps0, D5!="NA") #remove NAs
ps0 <- filter_taxa(ps0, function (x) {sum(x > 0) > 1}, prune=TRUE)
ps0 = subset_taxa(ps0, !D1 %in% c("D_1__Cyanobacteria"))
ps0
```
  
  
Visualize Prefiltered Data
```{r Prefilter Results}
#generate a prevelance table (number of samples each taxa occurs in) for each taxa using the following code:
prevdf = apply(X = otu_table(ps0),
               MARGIN = ifelse(taxa_are_rows(ps0), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
#Add taxonomy and total read counts to this data frame:
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps0),
                    tax_table(ps0))

#Phylum Level
plyr::ddply(prevdf, "D1", function(df1){
  data.frame(mean_prevalence=mean(df1$Prevalence),total_abundance=sum(df1$TotalAbundance,na.rm = T),stringsAsFactors = F)
  })

#Genus Level
plyr::ddply(prevdf, "D5", function(df1){
  data.frame(mean_prevalence=mean(df1$Prevalence),total_abundance=sum(df1$TotalAbundance,na.rm = T),stringsAsFactors = F)
  })
```
  
  
```{r Prefilter Plot}
###Plot 0% filtered data and visualize a 5%-20% threshold (Genus Level)
FilterPlot0 <- ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(ps0),color=D5)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  
  geom_hline(yintercept = 0.1, alpha = 0.5, linetype = 2) +  
  geom_hline(yintercept = 0.15, alpha = 0.5, linetype = 2) +  
  geom_hline(yintercept = 0.2, alpha = 0.5, linetype = 2) +  
  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~D1) + theme(legend.position="none")
ggsave("GenusLevelFilterPlot0.pdf", FilterPlot0, width = 30, height = 30, units = "in")
FilterPlot0
```
  
  
###Filtering
Investigate low prevelance/abundance genera and subset them out.
Define prevalence threshold
```{r Filtering}
prevalenceThreshold = 0.1 * nsamples(ps0)
prevalenceThreshold
```
  
  
Subset the remaining genera by prevalence
```{r Filtering and Subsetting by Prevalence}
keepTaxa = rownames(prevdf)[(prevdf$Prevalence >= prevalenceThreshold)]
length(keepTaxa)

psF = prune_taxa(keepTaxa, ps0)
psF
```
  
Compositional Plot 
```{r Compositional Plot}
psmeltF <- psmelt(psF)

abundantClasses <- psmeltF %>% group_by(D5) %>% summarize(frac = n() / nrow(.)) %>% na.omit() %>% pull(D5)

psmeltF %>% filter(D5 %in% abundantClasses) %>% ggplot(aes_string(x = "Generation", y = "Abundance", fill = "D5")) + geom_bar(stat = "identity", position = "fill") + theme(axis.text.x = element_text(angle = 0, hjust = 0), legend.text=element_text(size=rel(1))) + scale_fill_discrete(name="Genera")  + scale_y_sqrt() + ylab(expression(sqrt("Fraction of Community"))) + xlab("Generation") + facet_grid(SampleType~Treatment, scales="free")

ggsave("Filtered Data Barplot (Genus Level).pdf", height = 7, width = 12)
```
  
  
Subset Data for Analyses:
```{r Subset Gut vs Media}
psF.gut <- tax_glom(subset_samples(psF, SampleType=="Gut"), taxrank = "D5") 
psF.media <- tax_glom(subset_samples(psF, SampleType=="Media"), taxrank = "D5")
```
  
  
Gut Only Compositional Plot (for reference)
```{r gut Only Compositional Plot}
psmeltF.gut <- psmelt(psF.gut)

abundantClasses.gut <- psmeltF.gut %>% group_by(D5) %>% summarize(frac = n() / nrow(.)) %>% na.omit() %>% pull(D5)

psmeltF.gut %>% filter(D5 %in% abundantClasses.gut) %>% ggplot(aes_string(x = "Generation", y = "Abundance", fill = "D5")) + geom_bar(stat = "identity", position = "fill") + theme(axis.text.x = element_text(angle = 0, hjust = 0), legend.text=element_text(size=rel(1))) + scale_fill_discrete(name="Genera")  + scale_y_sqrt() + ylab(expression(sqrt("Fraction of Community"))) + xlab("Generation") + facet_grid(~Treatment, scales="free")

ggsave("Filtered Gut Barplot (Genus Level).pdf", height = 5, width = 12)
```
  
  
Media Only Compositional Plot (for reference)
```{r Media Only Compositional Plot}
psmeltF.media <- psmelt(psF.media)

abundantClasses.media <- psmeltF.media %>% group_by(D5) %>% summarize(frac = n() / nrow(.)) %>% na.omit() %>% pull(D5)

psmeltF.media %>% filter(D5 %in% abundantClasses.media) %>% ggplot(aes_string(x = "Generation", y = "Abundance", fill = "D5")) + geom_bar(stat = "identity", position = "fill") + theme(axis.text.x = element_text(angle = 0, hjust = 0), legend.text=element_text(size=rel(1))) + scale_fill_discrete(name="Genera")  + scale_y_sqrt() + ylab(expression(sqrt("Fraction of Community"))) + xlab("Generation") + facet_grid(~Treatment, scales="free")

ggsave("Filtered Media Barplot (Genus Level).pdf", height = 5, width = 12)
```
  
  
##Alpha diversity
Gut Samples:
```{r Gut Alpha Diversity}
divnet_gut <-  divnet(tax_glom(psF.gut, taxrank = "D5"), X = "GenTreat")

metatable1 <- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16s%20rRNA%20HiSeq%20Analysis/Metatable.txt")
colnames(metatable1)[1] <- "sample_names"

divnet_gut_summary <- divnet_gut$shannon %>% summary %>% left_join(metatable1) %>% group_by(TypeGenTreat) %>% summarize(alpha = estimate[1], lower = lower[1], upper = upper[1], SampleType = SampleType[1], Generation = Generation[1], Treatment = Treatment[1]) 

divnet_gut_summary %>% ggplot(aes(Treatment, alpha, color = Treatment)) + geom_point() + geom_linerange(aes(ymin = lower, ymax = upper)) + facet_grid(Generation~Treatment, scales = "free") + theme_minimal() + theme(axis.text.x=element_blank()) + ggtitle("Alpha Diversity (Gut Samples)") + xlab("Treatment") + ylab("Shannon Diversity (alpha)") + guides(color = F)

ggsave("Alpha Diversity (gut).pdf", height = 10, width = 7)
```
  
  
Media Samples:
```{r Media Alpha Diversity}
divnet_media <-  divnet(tax_glom(psF.media, taxrank = "D5"), X = "GenTreat")

metatable1 <- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16s%20rRNA%20HiSeq%20Analysis/Metatable.txt")
colnames(metatable1)[1] <- "sample_names"

divnet_media_summary <- divnet_media$shannon %>% summary %>% left_join(metatable1) %>% group_by(TypeGenTreat) %>% summarize(alpha = estimate[1], lower = lower[1], upper = upper[1], SampleType = SampleType[1], Generation = Generation[1], Treatment = Treatment[1]) 

divnet_media_summary %>% ggplot(aes(Treatment, alpha, color = Treatment)) + geom_point() + geom_linerange(aes(ymin = lower, ymax = upper)) + facet_grid(Generation~Treatment, scales = "free") + ggtitle("Alpha Diversity (Media)") + theme_minimal() + theme(axis.text.x=element_blank()) + xlab("Treatment") + ylab("Shannon Diversity (alpha)") + guides(color = F)

ggsave("Alpha Diversity (media).pdf", height = 10, width = 7)
```
  
  
  
##Examine detected disperson differences
###Using DESeq2
Due to plate set up during processing of microbiome genes for sequencing, F0/F1 has been separated from F2/F3 to help prevent potentially biasing batch effect analysis errors.
```{r DeSeq2 Gut Analysis}
#F0/F1 Data
psF.gut01 <- subset_samples(psF.gut, Generation %in% c("F0", "F1"))
ddse01 <- phyloseq_to_deseq2(psF.gut01, ~GenTreat+0)
ddse01 <- estimateSizeFactors(ddse01)
idx <- rowSums(counts(ddse01, normalized=TRUE) >= 5 ) >= 5 # additional abundance-based filtering, must have more than five counts in more than five samples in the design
ddse1 <- DESeq(ddse01[idx,], test="Wald", fitType="local")
resultsNames(ddse1)
#Contrasts:
res0 <- results(ddse1, contrast=c("GenTreat","F0NSD","F0CD")) #numerator (control) , denominator(nsd)
res1 <- results(ddse1, contrast=c("GenTreat","F1NSD","F1CD"))

#F2/F3 Data
psF.gut23 <- subset_samples(psF.gut, Generation %in% c("F2", "F3"))
ddse23 <- phyloseq_to_deseq2(psF.gut23, ~GenTreat+0)
ddse23 <- estimateSizeFactors(ddse23)
idx <- rowSums(counts(ddse23, normalized=TRUE) >= 5 ) >= 5 # additional abundance-based filtering, must have more than five counts in more than five samples in the design
ddse2 <- DESeq(ddse23[idx,], test="Wald", fitType="local")
resultsNames(ddse2)
#Contrasts:
res2 <- results(ddse2, contrast=c("GenTreat","F2NSD","F2CD")) 
res3 <- results(ddse2, contrast=c("GenTreat","F3NSD","F3CD"))

table(res0$padj < 0.05)
table(res1$padj < 0.05)
table(res2$padj < 0.05)
table(res3$padj < 0.05)

res0$genus <- gsub("D_5__", "", tax_table(psF.gut)[rownames(res0)])
res1$genus <- gsub("D_5__", "", tax_table(psF.gut)[rownames(res1)])
res2$genus <- gsub("D_5__", "", tax_table(psF.gut)[rownames(res2)])
res3$genus <- gsub("D_5__", "", tax_table(psF.gut)[rownames(res3)])

bind_rows(as.data.frame(res0)[,c("genus.D5", "log2FoldChange", "padj")],
          as.data.frame(res1)[,c("genus.D5", "log2FoldChange", "padj")],
          as.data.frame(res2)[,c("genus.D5", "log2FoldChange", "padj")],
          as.data.frame(res3)[,c("genus.D5", "log2FoldChange", "padj")],
          .id = "generations") %>% 
  mutate(generations = as.numeric(generations) - 1, sig = ifelse(padj<0.05, 1, .5)) %>%
  ggplot(aes(generations, log2FoldChange, color = genus.D5, alpha = sig, group = genus.D5)) + geom_point(aes(size=5)) + geom_line(alpha=0.5) + guides(size = "none", alpha = "none")

```
  
  
Table and Plot Significant Results
```{r}
phenotypes <- read_csv("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/Phenotype%20Analysis/phenotype_data") %>% filter(! grepl("\\+", Treatment) & Sex == "female") %>% mutate(Treatment = factor(Treatment, labels = c("CD", "NSD"))) %>% left_join(psF.gut@sam_data, ., by = c("Generation", "Treatment", "SampleNumber" = "Sample")) %>% select(Plate, Generation, Treatment, GenTreat, TransGen, Fat.Weight, TAG) %>% mutate (TAGwt = TAG/Fat.Weight)

psF.gut@sam_data$TAGwt <- phenotypes$TAGwt
psF.gutMet <- subset_samples(psF.gut,  Generation %in% c("F1", "F2", "F3"))
ddseMet <- phyloseq_to_deseq2(psF.gutMet, ~ TAGwt )
ddseMet <- estimateSizeFactors(ddseMet)
idx <- rowSums(counts(ddseMet, normalized=TRUE) >= 5 ) >= 5
ddseMet2 <- DESeq(ddseMet[idx,], test="Wald", fitType="local")
resultsNames(ddseMet2)
table(results(ddseMet2, name="TAGwt")$padj<0.05)
(tagsig <- results(ddseMet2, name="TAGwt")[results(ddseMet2, name="TAGwt")$padj<0.05,])
tax_table(psF.gut)[rownames(tagsig),6]
metCounts <- data.table::rbindlist(lapply(1:3, function(x) {  plotCounts(ddseMet, gene=rownames(tagsig)[x], intgroup="TAGwt",returnData=TRUE) %>% mutate(taxon = as.vector(gsub("D_5__","", tax_table(psF.gut)[rownames(tagsig)[x],6])), Generation = colData(ddseMet2)$Generation, Treatment = colData(ddseMet2)$Treatment) })) 
metCounts %>% ggplot(aes(count, TAGwt, color=Generation, shape=Treatment ))+geom_point()+scale_x_log10(limit=c(1,NA))+facet_grid(taxon~.)+stat_smooth(method="lm",aes(group=0))  
```

###Using Permutation Test
```{r Permutation Analyses Gut}
#F0
psF.gut.F0 <- tax_glom(subset_samples(psF.gut, Generation=="F0"), taxrank = "D5")  
gut.F0 = UniFrac(psF.gut.F0)
#tests homogeneity of dispersion among groups (https://www.researchgate.net/post/Betadisper_and_adonis_in_R_Am_I_interpreting_my_output_correctly)
gut.F0.adonis <- adonis2(gut.F0 ~ Treatment, as(sample_data(psF.gut.F0), "data.frame"), permutations = 10000, strata = line)
gut.F0.adonis
#tests for differences between groups
anova(betadisper(gut.F0, sample_data(psF.gut.F0)$Treatment))

#F1
psF.gut.F1 <- tax_glom(subset_samples(psF.gut, Generation=="F1"), taxrank = "D5")  
gut.F1 = UniFrac(psF.gut.F1)
#tests homogeneity of dispersion among groups
gut.F1.adonis <- adonis2(gut.F1 ~ Treatment, as(sample_data(psF.gut.F1), "data.frame"), permutations = 10000, strata = line)
gut.F1.adonis
#tests for differences between groups
anova(betadisper(gut.F1, sample_data(psF.gut.F1)$Treatment))

#F2
psF.gut.F2 <- tax_glom(subset_samples(psF.gut, Generation=="F2"), taxrank = "D5")  
gut.F2 = UniFrac(psF.gut.F2)
#tests homogeneity of dispersion among groups (https://www.researchgate.net/post/Betadisper_and_adonis_in_R_Am_I_interpreting_my_output_correctly)
gut.F2.adonis <- adonis2(gut.F2 ~ Treatment, as(sample_data(psF.gut.F2), "data.frame"), permutations = 10000, strata = line)
gut.F2.adonis
#tests for differences between groups
anova(betadisper(gut.F2, sample_data(psF.gut.F2)$Treatment))

#F3
psF.gut.F3 <- tax_glom(subset_samples(psF.gut, Generation=="F3"), taxrank = "D5")  
gut.F3 = UniFrac(psF.gut.F3)
#tests homogeneity of dispersion among groups (https://www.researchgate.net/post/Betadisper_and_adonis_in_R_Am_I_interpreting_my_output_correctly)
gut.F3.adonis <- adonis2(gut.F3 ~ Treatment, as(sample_data(psF.gut.F3), "data.frame"), permutations = 10000, strata = line)
gut.F3.adonis
#tests for differences between groups
anova(betadisper(gut.F3, sample_data(psF.gut.F3)$Treatment))
```
**Using the permutation test gives no significant results. Given that DESeq2 is considered the more reliable and sensitive method according to several publications, I think we remove the permutational tests from the analyses all together. 


###Media:
```{r DeSeq2 Media Analysis}
#F0/F1 Data
psF.media01 <- subset_samples(psF.media, Generation %in% c("F0", "F1"))
ddseM01 <- phyloseq_to_deseq2(psF.media01, ~GenTreat+0)
ddseM01 <- estimateSizeFactors(ddseM01)
ddseM01 <- estimateDispersionsGeneEst(ddseM01)
dispersions(ddseM01) <- mcols(ddseM01)$dispGeneEst
idx <- rowSums(counts(ddseM01, normalized=TRUE) >= 5 ) >= 5 # additional abundance-based filtering, must have more than five counts in more than five samples in the design
ddseM1 <- DESeq(ddseM01[idx,], test="Wald", fitType="local")
resultsNames(ddseM1)
#Contrasts:
res0M <- results(ddseM1, contrast=c("GenTreat","F0CD","F0NSD"))
res1M <- results(ddseM1, contrast=c("GenTreat","F1CD","F1NSD"))

#F2/F3 Data
psF.media23 <- subset_samples(psF.media, Generation %in% c("F2", "F3"))
ddseM23 <- phyloseq_to_deseq2(psF.media23, ~GenTreat+0)
ddseM23 <- estimateSizeFactors(ddseM23)
idx <- rowSums(counts(ddseM23, normalized=TRUE) >= 5 ) >= 5 # additional abundance-based filtering, must have more than five counts in more than five samples in the design
ddseM2 <- DESeq(ddseM23[idx,], test="Wald", fitType="local")
resultsNames(ddseM2)
#Contrasts:
resM2 <- results(ddseM2, contrast=c("GenTreat","F2CD","F2NSD"))
resM3 <- results(ddseM2, contrast=c("GenTreat","F3CD","F3NSD"))

table(resM0$padj < 0.05)
table(resM1$padj < 0.05)
table(resM2$padj < 0.05)
table(resM3$padj < 0.05)
```


#### Media permutation tests
```{r}
psF.media.F0 <- tax_glom(subset_samples(psF.media, Generation=="F0"), taxrank = "D5")  
media.F0 = UniFrac(psF.media.F0)
#tests homogeneity of dispersion among groups (https://www.researchgate.net/post/Betadisper_and_adonis_in_R_Am_I_interpreting_my_output_correctly)
media.F0.adonis <- adonis2(media.F0 ~ Treatment, as(sample_data(psF.media.F0), "data.frame"), permutations = 10000, strata = line)
media.F0.adonis
#tests for differences between groups
anova(betadisper(media.F0, sample_data(psF.media.F0)$Treatment))


psF.media.F1 <- tax_glom(subset_samples(psF.media, Generation=="F1"), taxrank = "D5")  
media.F1 = UniFrac(psF.media.F1)
#tests homogeneity of dispersion among groups
media.F1.adonis <- adonis2(media.F1 ~ Treatment, as(sample_data(psF.media.F1), "data.frame"), permutations = 10000, strata = line)
media.F1.adonis
#tests for differences between groups
anova(betadisper(media.F1, sample_data(psF.media.F1)$Treatment))

psF.media.F2 <- tax_glom(subset_samples(psF.media, Generation= "F2"), taxrank = "D5")  
media.F2 = UniFrac(psF.media.F2)
#tests homogeneity of dispersion among groups (https://www.researchgate.net/post/Betadisper_and_adonis_in_R_Am_I_interpreting_my_output_correctly)
media.F2.adonis <- adonis2(media.F2 ~ Treatment, as(sample_data(psF.media.F2), "data.frame"), permutations = 10000, strata = line)
media.F2.adonis
#tests for differences between groups
anova(betadisper(media.F2, sample_data(psF.media.F2)$Treatment))
plot(betadisper(media.F2, sample_data(psF.media.F2)$Treatment))

psF.media.F3 <- tax_glom(subset_samples(psF.media, Generation= "F3"), taxrank = "D5")  
media.F3 = UniFrac(psF.media.F3)
#tests homogeneity of dispersion among groups (https://www.researchgate.net/post/Betadisper_and_adonis_in_R_Am_I_interpreting_my_output_correctly)
media.F3.adonis <- adonis2(media.F3 ~ Treatment, as(sample_data(psF.media.F3), "data.frame"), permutations = 10000, strata = line)
media.F3.adonis
#tests for differences between groups
anova(betadisper(media.F3, sample_data(psF.media.F3)$Treatment))
plot(betadisper(media.F3, sample_data(psF.media.F3)$Treatment))
```
