---
title: "16s rRNA Analysis"
author: "Carmen Emborski"
date: "5/9/2018"
output: html_document
---
Packages
```{r setup, include=FALSE}
theme_set(theme_bw())

library("phyloseq")
library("ggplot2")
library("dplyr")
library("tidyr")
library("RColorBrewer")
library("gridExtra")
library("vegan")
library("DESeq2")
```
Import Files
```{r}
taxonomy<- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16S_rRNA_analysis/Taxonomy.tsv")
names(taxonomy) <- c("row", "tax", "Confidence")
row.names(taxonomy) <-taxonomy[[1]]
taxonomy <- taxonomy[,(-1)]
taxonomy <-  separate(taxonomy, tax, c("D0","D1", "D2", "D3", "D4", "D5", "D6", "D7", "D8", "D9", "D10", "D11", "D12", "D13", "D14"), sep = ";", fill = "right")
taxonomy <- taxonomy[,c(1:6)]
taxmat <- as.matrix(taxonomy)
TAX = tax_table(taxmat) 

#import metadata
metatable <- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16S_rRNA_analysis/Metatable.txt")
row.names(metatable) <- metatable[[1]]
metatable<- metatable[,(-1)]
META <- sample_data(metatable)

#Import Phylogenetic tree (rooted, exported from qiime2)
TREE <- read_tree("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16S_rRNA_analysis/tree.nwk")

#Import feature (sequence variant) table
Svtab <- read.delim("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/16S_rRNA_analysis/FeatureTable.txt")
row.names(Svtab)<-Svtab[[1]] #make OTU ID the row names
Svtab<-Svtab[,-(1)] # remove OTU ID column
fmat <- as.matrix(Svtab)
OTU = otu_table(fmat, taxa_are_rows = TRUE)

#create phyloseq object
ps <- phyloseq(OTU, TAX, META, TREE)
```
Filter Data (Prevalence Filtering)
```{r}
#Filter Data
#0% THRESHOLD
rank_names(ps)
ps <- subset_taxa(ps, D0=="D_0__Bacteria") #keep only bacterial ASVs
table(tax_table(ps)[, "D0"], exclude = NULL)
ps = subset_taxa(ps, D1!="NA") #remove NAs

# function to remove rare artifactual taxa
# thresh is the fraction of samples that must have these OTUs
filterTaxa <- function(ps, thresh) {
  table(tax_table(ps)[, "D0"], exclude = NULL) # remove sequences that have D0 unassigned
  ps <- subset_taxa(ps, D0=="D_0__Bacteria") # Filter Physeq object - Bacteria ONLY
  table(tax_table(ps)[, "D2"], exclude = NULL)
  ps = subset_taxa(ps, !D2 %in% c("D_2__Chloroplast" , " D_2__Cyanobacteria")) #remove phyla that cannot be part of microbiome
  
  prevdf = apply(X = otu_table(ps),
                 MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
                 FUN = function(x){sum(x > 0)})
  
  prevdf = data.frame(Prevalence = prevdf,
                      TotalAbundance = taxa_sums(ps),
                      tax_table(ps))
  
  plyr::ddply(prevdf, "D2", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
  
  prevdf1.plot <- ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(ps),color=D2)) +
    # Include a guess for parameter
    geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
    scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
    facet_wrap(~D2) + theme(legend.position="none")
  
  prevalenceThreshold = thresh * nsamples(ps) # remove OTS found in less than thres% of the samples
  
  keepTaxa = rownames(prevdf)[(prevdf$Prevalence >= prevalenceThreshold)]
  ps2 = prune_taxa(keepTaxa, ps)
  
  table(tax_table(ps2)[, "D2"], exclude = NULL)
  
  prevdf2 = apply(X = otu_table(ps2),
                 MARGIN = ifelse(taxa_are_rows(ps2), yes = 1, no = 2),
                 FUN = function(x){sum(x > 0)})
  
  prevdf2 = data.frame(Prevalence = prevdf2,
                      TotalAbundance = taxa_sums(ps2),
                      tax_table(ps2))
  
  plyr::ddply(prevdf2, "D1", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
  
  #transform to relative abundance
  ps2ra = transform_sample_counts(ps2, function(x){x / sum(x)})
  
  ##################################
  #Remove all taxa that are D1 == NA
  
  psF = subset_taxa(ps2, D1 %in% c("D_1__Acidobacteria", "D_1__Actinobacteria", "D_1__Bacteroidetes", "D_1__Firmicutes", " D_1__Parcubacteria", "D_1__Proteobacteria", "D_1__Verrucomicrobia"))
  table(tax_table(psF)[, "D1"], exclude = NULL)
  
  #Final Prevalence Plot
  prevdf3 = apply(X = otu_table(psF),
                  MARGIN = ifelse(taxa_are_rows(psF), yes = 1, no = 2),
                  FUN = function(x){sum(x > 0)})
  
  prevdf3 = data.frame(Prevalence = prevdf3,
                       TotalAbundance = taxa_sums(psF),
                       tax_table(psF))
  
  plyr::ddply(prevdf3, "D1", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
  
  ggplot(prevdf3, aes(TotalAbundance, Prevalence / nsamples(ps),color=D2)) +
    # Include a guess for parameter
    geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
    scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
    facet_wrap(~D2) + theme(legend.position="none")
  return(psF)
}
```
Analyze for quantitative changes in individual species (ASVs)
```{r}
psF <- filterTaxa(ps, 0.20) #abundance based filtering - keep only species present in 20% of samples

ps01 <- subset_samples(psF, Generation %in% c("F0", "F1"))
ddse01 <- phyloseq_to_deseq2(ps01, ~ GenTreat)
ddse01 <- estimateSizeFactors(ddse01)
idx <- rowSums(counts(ddse01, normalized=TRUE) >= 5 ) >= 5 # additional abundance-based filtering, must have more than five counts in more than five samples in the design
ddse01.2 <- DESeq(ddse01[idx,], test="Wald", fitType="local", betaPrior=FALSE)

resultsNames(ddse01.2)

res0 <- results(ddse01.2, contrast=c("GenTreat","F0C","F0N"))
table(res0$padj < 0.05)
res0$species <- gsub("D_8__", "", tax_table(psF)[rownames(res0)])
res0[res0$padj < 0.05,c("species", "log2FoldChange", "padj")]
#Interestingly all F0 N ASVs that significantly differed from C were downregulated

res1 <- results(ddse01.2, contrast=c("GenTreat","F1C","F1N"))
table(res1$padj < 0.05)
res1$species <- gsub("D_6__","", tax_table(psF)[rownames(res1)])
res1[res1$padj < 0.05,c("species", "log2FoldChange", "padj")]
res1vres0 <- res0[rownames(res1)[which(res1$padj<0.05)], c("species", "log2FoldChange", "padj")]
res1vres0
#Interestingly, all F1 N ASVs that significantly differed from C were upregulated, and both species are found in F0

#Due to possible (study design) batch effects, F2/F3 were run in their own model
psF23 <- subset_samples(psF, Generation %in% c("F2", "F3"))
ddse23 <- phyloseq_to_deseq2(psF23, ~ GenTreat)
ddse23 <- estimateSizeFactors(ddse23)
idx <- rowSums(counts(ddse23, normalized=TRUE) >= 5 ) >= 5 # additional abundance-based filtering, must have more than five counts in more than five samples in the design
ddse23.2 <- DESeq(ddse23[idx,], test="Wald", fitType="local", betaPrior=FALSE)

res2 <- results(ddse23.2, contrast=c("GenTreat","F2C","F2N"))
table(res2$padj < 0.05)
res2$species <- gsub("D_5__", "", tax_table(psF)[rownames(res2)])
res2[res2$padj < 0.05,c("species", "log2FoldChange", "padj")]
res2vres0 <- res0[rownames(res2)[which(res2$padj<0.05)], c("species", "log2FoldChange", "padj")]
#res2vres0
#Although no similar significant ASVs were found between F0 and F2, there were 4 F2 N ASVs (species) that were downregulated relative to controls

res3 <- results(ddse23.2, contrast=c("GenTreat","F3C","F3N"))
table(res3$padj < 0.05)
#No F3 N ASVs (species) significantly differed from controls

#write.csv(res0, file = "res0.csv")
#write.csv(res1, file = "res1.csv")
#write.csv(res2, file = "res2.csv")
#write.csv(res3, file = "res3.csv")


# returns data frame of normalized counts for plotting from a deseq object and a table of results
combineSig <- function(ddse01, res) {
  sigtab <- res[which(res$padj < 0.05), ]
  taxCounts <- list()
  i <- 1
  for (taxon in rownames(sigtab)) {
    taxName <- as.data.frame(tax_table(ps01)[taxon])[,6]
    tc <- plotCounts(ddse01.2, gene=taxon, intgroup="GenTreat",returnData=TRUE)
    tc$species <- taxName
    tc$code <- paste(taxon,taxName)
    taxCounts[[i]] <- tc %>% separate(GenTreat, into = c("Gen", "Treat"), sep=2) %>% mutate(species = gsub("D_5__", "", species))
    i <- i + 1
  }
  return(plyr::rbind.fill(taxCounts))
}

#PLOT F0/F1
fig3a <- combineSig(ddse01.2, res1) %>% ggplot(aes(Treat,count,fill=Treat)) + geom_jitter(width=.1, pch=21, colour="black", size=1.75) + ggtitle("Horizontally Transmitted Bacteria") + xlab("Generations") + ylab("ASV count") + scale_fill_manual("Non-Antibiotic\nTreatment", values=c("#FF9966", "#33CCFF"))  + facet_grid(species~Gen)+scale_y_log10() + theme_bw() + theme(plot.title = element_text(hjust = 0.5))

ggsave("fig3a.png", fig3a, width = 6, height = 5, units = "in")




#Are any bacteria associated with TAGs?
phenotypes <- read.csv("https://raw.githubusercontent.com/cemborski/Horizontal-transmission-of-altered-phenotypes/master/Phenotype%20Analysis/phenotype_data") %>% filter(!grepl("\\+", Treatment)) %>% mutate(Treatment = factor(Treatment, labels = c("C", "N"))) %>% left_join(psF@sam_data, ., by = c("Generation", "Treatment", "SampleNumber" = "Sample")) %>% mutate (TAGwt = TAG/Fat.Weight)

psF@sam_data$TAGwt <- phenotypes$TAGwt

psFMet <- subset_samples(psF,  Generation %in% c("F1", "F2", "F3"))
ddseMet <- phyloseq_to_deseq2(psFMet, ~ TAGwt)
ddseMet <- estimateSizeFactors(ddseMet)
idx <- rowSums(counts(ddseMet, normalized=TRUE) >= 5 ) >= 5
ddseMet2 <- DESeq(ddseMet[idx,], test="Wald", fitType="local", betaPrior=FALSE)
resultsNames(ddseMet2)
table(results(ddseMet2, name="TAGwt")$padj<0.05)
(tagsig <- results(ddseMet2, name="TAGwt")[results(ddseMet2, name="TAGwt")$padj<0.05,])
tax_table(psF)[rownames(tagsig),6]

metCounts <- data.table::rbindlist(lapply(1:3, function(x) {  plotCounts(ddseMet, gene=rownames(tagsig)[x], intgroup="TAGwt",returnData=TRUE) %>% mutate(taxon = as.vector(gsub("D_5__","", tax_table(psF)[rownames(tagsig)[x],6])), Generation = colData(ddseMet2)$Generation, Treatment = colData(ddseMet2)$Treatment) })) 

TAGvASV.plot <- metCounts %>% ggplot(aes(count, TAGwt, color=Treatment, shape=Generation )) + geom_point() + scale_x_log10(limit=c(1,NA)) + facet_grid(taxon~.) + scale_color_manual("Treatment", values=c("#FF9966", "#33CCFF")) + ggtitle("Relationship between bacteria and phenotype") + xlab("Abundance") + ylab("Triglycerides/Fly Weight") + stat_smooth(method="lm",aes(group=0)) + theme_bw() + theme(plot.title = element_text(hjust = 0.5))

ggsave("TAGvASVplot.png", TAGvASV.plot, width = 5, height = 5, units = "in")


Figure3.plot <- grid.arrange(arrangeGrob(fig3a + theme(legend.position="none"), TAGvASV.plot, ncol=2))

ggsave("Figure3.png", Figure3.plot, width = 10, height = 5, units = "in")



```

